File: Excel/612 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/612 Names Having First and Last Names 1st Characters Same.xlsx"
input = read_excel(path, range = "A1:A20")
test  = read_excel(path, range = "B1:B6")

result = input %>%
  mutate(initials = str_extract_all(Names, "[A-Z]") %>% 
           map_chr(~ paste0(.[1], .[length(.)]))) %>%
  filter(n() > 1, .by = initials) %>%
  select(Names) %>%
  arrange(Names)

all.equal(result$Names, test$`Answer Expected`)
# [1] TRUE

File: Excel/600-699/612/612 Challenge.py
import pandas as pd
import re

path = "612 Names Having First and Last Names 1st Characters Same.xlsx"
input = pd.read_excel(path, usecols="A", nrows=20)
test = pd.read_excel(path, usecols="B", nrows=5)

extract_initials = lambda name: ''.join(re.findall(r'[A-Z]', name)[:1] + re.findall(r'[A-Z]', name)[-1:])
result = input.assign(initials=input['Names']\
                .apply(extract_initials))\
                .groupby('initials')\
                .filter(lambda x: len(x) > 1)\
                .sort_values(by='Names').reset_index()

print(result['Names'].equals(test['Answer Expected'])) # True